node/src/exchange/server.rs

```diff
@@ -1,52 +1,396 @@
+use std::fmt::{Debug, Display};
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 
-use celestia_proto::p2p::pb::{HeaderRequest, HeaderResponse};
+use celestia_proto::p2p::pb::{header_request, HeaderRequest, HeaderResponse};
+use celestia_types::Hash;
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, Stream};
 use libp2p::{
     request_response::{InboundFailure, RequestId, ResponseChannel},
     PeerId,
 };
-use tracing::instrument;
+use tendermint::hash::Algorithm;
+use tracing::{instrument, trace};
 
+use crate::exchange::utils::{ExtendedHeaderExt, HeaderRequestExt, HeaderResponseExt};
+use crate::exchange::{ReqRespBehaviour, ResponseType};
 use crate::store::Store;
 
-pub(super) struct ExchangeServerHandler<S>
+type StoreJobType<C> = dyn Future<Output = (C, ResponseType)> + Send;
+
+pub(super) struct ExchangeServerHandler<S, R = ReqRespBehaviour>
 where
-    S: Store + 'static,
+    S: Store,
+    R: ResponseSender,
 {
-    _store: Arc<S>,
+    store: Arc<S>,
+    store_jobs: FuturesUnordered<Pin<Box<StoreJobType<R::Channel>>>>,
```

Comment:
As I said in another comment we can use [`BoxFuture`](https://docs.rs/futures/latest/futures/future/type.BoxFuture.html):

```suggestion
    store_jobs: FuturesUnordered<BoxFuture<(R::Channel, ResponseType)>>,
```

Reply:
switched to BoxFuture and `futures` 👍